Adopt more smart pointers in WebScreenOrientationManagerProxy and UserMediaCaptureManagerProxy

RupinMittal opened the pull request:
#### c29543ecbb6ad433a92582c629fd2828a8912447
<pre>
Adopt more smart pointers in WebScreenOrientationManagerProxy and UserMediaCaptureManagerProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=279952">https://bugs.webkit.org/show_bug.cgi?id=279952</a>
<a href="https://rdar.apple.com/136268991">rdar://136268991</a>

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createCameraSource):
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):
(WebKit::UserMediaCaptureManagerProxy::capabilities):
(WebKit::UserMediaCaptureManagerProxy::applyConstraints):
(WebKit::UserMediaCaptureManagerProxy::clone):
(WebKit::UserMediaCaptureManagerProxy::takePhoto):
(WebKit::UserMediaCaptureManagerProxy::getPhotoCapabilities):
(WebKit::UserMediaCaptureManagerProxy::getPhotoSettings):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::protectedConnection):
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::protectedLogger):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::~WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::setCurrentOrientation):
(WebKit::WebScreenOrientationManagerProxy::protectedPage const):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.h:

Canonical link: <a href="https://commits.webkit.org/283959@main">https://commits.webkit.org/283959@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecbca83ff99d4e5357a3b53957457c3375a86614

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67873 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47259 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71929 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19014 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55057 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18830 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54294 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12703 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43305 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58677 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34751 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39980 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17372 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61944 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16432 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73626 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11840 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15716 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61737 "Build was cancelled. Recent messages:Printed configuration; Skipped layout-tests; 55 flakes 128 failures; Uploaded test results; 67 flakes 50 failures; Compiled WebKit (cancelled)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11877 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58753 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9644 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3255 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10335 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43064 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45330 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43880 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
```diff
@@ -190,10 +190,12 @@ class GPUProxyForCapture final : public UserMediaCaptureManagerProxy::Connection
 
 private:
-    Logger& logger() final { return m_process.get()->logger(); }
+    Logger& logger() const final { return m_process.get()->logger(); }
```
Why are you making these functions const?

RupinMittal replied:
I was adding a `protectedLogger()` function right next to `logger()` and saw that `logger()` doesn't modify anything and could be marked const--so I thought I'd add it. Also I see that many `protectedFoo()` functions are const so I thought I'd made this one const if possible, which meant making `logger()` const as well.

RupinMittal replied:
Actually this const is wrong because calling `logger` does modify the `m_process` member variable. I'll remove this const.

RupinMittal replied:
Fixed!

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -185,7 +186,7 @@ class UserMediaCaptureManagerProxySourceProxy final
         ASSERT(isMainRunLoop());
 
-        m_pendingAction = m_pendingAction->isResolved() ? action() : m_pendingAction->whenSettled(RunLoop::main(), WTFMove(action));
+        m_pendingAction = protectedPendingAction()->isResolved() ? action() : m_pendingAction->whenSettled(RunLoop::protectedMain(), WTFMove(action));
```
Let's' take a:
RefPtr protectedPendingAction = this->protectedPendingAction().
And use it for both isResolved and whenSettled.

RupinMittal replied:
Fixed!

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -232,9 +233,9 @@ class UserMediaCaptureManagerProxySourceProxy final
             m_settings = m_source->settings();
             if (m_widthConstraint || m_heightConstraint) {
-                auto desiredSize = m_source->computeResizedVideoFrameSize({ m_widthConstraint, m_heightConstraint }, m_source->intrinsicSize());
+                auto desiredSize = m_source->computeResizedVideoFrameSize({ m_widthConstraint, m_heightConstraint }, protectedSource()->intrinsicSize());
```
Ditto here, let's take protectedSource local variable for `m_source->computeResizedVideoFrameSize`, `intrinscSize` and `videoFrameRotation`

RupinMittal replied:
Fixed!

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -270,7 +271,7 @@ class UserMediaCaptureManagerProxySourceProxy final
             }
 
-            return m_source->takePhoto(WTFMove(photoSettings))->whenSettled(RunLoop::main(), [takePhotoProducer = WTFMove(takePhotoProducer)] (auto&& result) mutable {
+            return m_source->takePhoto(WTFMove(photoSettings))->whenSettled(RunLoop::protectedMain(), [takePhotoProducer = WTFMove(takePhotoProducer)] (auto&& result) mutable {
```
m_source?

RupinMittal replied:
Good catch--fixed!

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -410,6 +411,9 @@ class UserMediaCaptureManagerProxySourceProxy final
     }
 
+    Ref<GenericPromise> protectedPendingAction() const { return m_pendingAction; }
+    Ref<RealtimeMediaSource> protectedSource() const { return m_source; }
```
Might be worth checking the coding safe rules, but I wonder if using const Ref<> m_source (m_source is only set in the constructor) would allow us to directly use m_source instead of localling using protectedSource()?

cdumez replied:
We need a smart pointer on the stack as per the rules so using a protected function is correct.

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
```diff
@@ -70,8 +70,10 @@ class UserMediaCaptureManagerProxy : public IPC::MessageReceiver {
         virtual IPC::Connection& connection() = 0;
+        virtual Ref<IPC::Connection> protectedConnection() = 0;
         virtual bool willStartCapture(WebCore::CaptureDevice::DeviceType) const = 0;
         virtual Logger& logger() = 0;
```
Can we remove this one?

cdumez replied:
We shouldn't. It is still useful when calling trivial functions which do not require a smart pointer.

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
```diff
@@ -70,8 +70,10 @@ class UserMediaCaptureManagerProxy : public IPC::MessageReceiver {
         virtual void addMessageReceiver(IPC::ReceiverName, IPC::MessageReceiver&) = 0;
         virtual void removeMessageReceiver(IPC::ReceiverName) = 0;
         virtual IPC::Connection& connection() = 0;
```
Can we remove this one?

cdumez replied:
We shouldn't. It is still useful when calling trivial functions which do not require a smart pointer.

youennf commented on Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
```diff
@@ -45,14 +45,14 @@ WebScreenOrientationManagerProxy::WebScreenOrientationManagerProxy(WebPageProxy&
     , m_currentOrientation(orientation)
 {
-    m_page->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebScreenOrientationManagerProxy::messageReceiverName(), m_page->webPageIDInMainFrameProcess(), *this);
+    protectedPage()->protectedLegacyMainFrameProcess()->addMessageReceiver(Messages::WebScreenOrientationManagerProxy::messageReceiverName(), m_page->webPageIDInMainFrameProcess(), *this);
```
We are using m_page for webPageIDInMainFrameProcess.
Should we use a protectedPage local variable?

cdumez replied:
No need, webPageIDInMainFrameProcess() is a trivial inline function so the static analyzer shouldn't complain about it.

cdumez commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -112,7 +112,8 @@ class UserMediaCaptureManagerProxySourceProxy final
     }
 
-    RealtimeMediaSource& source() { return m_source; }
+    RealtimeMediaSource& source() const { return m_source; }
```
Why add const?

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -112,7 +112,8 @@ class UserMediaCaptureManagerProxySourceProxy final
 
-    RealtimeMediaSource& source() { return m_source; }
+    RealtimeMediaSource& source() const { return m_source; }
+    Ref<RealtimeMediaSource> protectedSource() const { return source(); }
```
Why add const?

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
```diff
@@ -70,8 +70,10 @@ class UserMediaCaptureManagerProxy : public IPC::MessageReceiver {
         virtual void addMessageReceiver(IPC::ReceiverName, IPC::MessageReceiver&) = 0;
         virtual void removeMessageReceiver(IPC::ReceiverName) = 0;
         virtual IPC::Connection& connection() = 0;
+        virtual Ref<IPC::Connection> protectedConnection() = 0;
```
This doesn't technically need to be virtual and overriden in subclasses, we could just do:
```
Ref<IPC::Connection> protectedConnection() { return connection(); }
```
here.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
```diff
@@ -70,8 +70,10 @@ class UserMediaCaptureManagerProxy : public IPC::MessageReceiver {
+        virtual Ref<IPC::Connection> protectedConnection() = 0;
         virtual bool willStartCapture(WebCore::CaptureDevice::DeviceType) const = 0;
         virtual Logger& logger() = 0;
+        virtual Ref<Logger> protectedLogger() = 0;
```
ditto.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -185,7 +186,8 @@ class UserMediaCaptureManagerProxySourceProxy final
         ASSERT(isMainRunLoop());
 
-        m_pendingAction = m_pendingAction->isResolved() ? action() : m_pendingAction->whenSettled(RunLoop::main(), WTFMove(action));
+        Ref pendingAction = m_pendingAction;
```
It would be a little more efficient to use:
```
Ref pendingAction = WTFMove(m_pendingAction);
```
We're about to reassign m_pendingAction on the next line so we don't care about leaving it in a good state and we can just steal it's "ref" to avoid refcounting churn.

RupinMittal replied:
Fixed!

cdumez reviewed (approved):
LGTM with nit.

RupinMittal committed:
Adopt more smart pointers in WebScreenOrientationManagerProxy and UserMediaCaptureManagerProxy
https://bugs.webkit.org/show_bug.cgi?id=279952
rdar://136268991

Reviewed by Chris Dumez.

Smart pointer adoption as per the static analyzer.

* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::createCameraSource):
(WebKit::UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstraints):
(WebKit::UserMediaCaptureManagerProxy::capabilities):
(WebKit::UserMediaCaptureManagerProxy::applyConstraints):
(WebKit::UserMediaCaptureManagerProxy::clone):
(WebKit::UserMediaCaptureManagerProxy::takePhoto):
(WebKit::UserMediaCaptureManagerProxy::getPhotoCapabilities):
(WebKit::UserMediaCaptureManagerProxy::getPhotoSettings):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::protectedConnection):
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::protectedLogger):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::~WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::setCurrentOrientation):
(WebKit::WebScreenOrientationManagerProxy::protectedPage const):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.h:

Canonical link: https://commits.webkit.org/283959@main
